cloudify_aws/s3/object.py

```diff
@@ -0,0 +1,171 @@
+
+# This package imports
+from cloudify_aws import constants
+from cloudify_aws.base import AwsBaseNode
+from cloudify_aws.connection import EC2ConnectionClient
+
+
+@operation
+def create(args=None, **_):
+    return Object().created(args)
+
+
+@operation
+def delete(args=None, **_):
+    return Object().deleted(args)
+
+
+class Object(AwsBaseNode):
+    def __init__(self, client=None):
+        client = client or EC2ConnectionClient().client3('s3')
+
+        ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID] = (
+            '{bucket}:{key}'  # Can only be populated after client exists
+        )
+        super(Object, self).__init__(
+            constants.S3_OBJECT['AWS_RESOURCE_TYPE'],
+            constants.S3_OBJECT['REQUIRED_PROPERTIES'],
+            client=client,
+        )
+
+        bucket_name, _ = self._get_bucket_details()
+
+        ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID] = (
+            '{bucket}:{key}'.format(
+                bucket=bucket_name,
+                key=ctx.node.properties['name'],
+            )
+        )
+
+        self.not_found_error = constants.S3_OBJECT['NOT_FOUND_ERROR']
+        self.get_all_handler = {
+            'function': self.client.list_objects,
+            'argument': 'Bucket',
+            'results_key': 'Contents',
+            'id_key': 'Key',
+        }
+
+    def _get_bucket_details(self):
+        relationships = ctx.instance.relationships
+
+        if len(relationships) == 0:
+            raise NonRecoverableError(
+                'S3 objects must be contained in S3 buckets using a '
+                'cloudify.aws.relationships.s3_object_contained_in_bucket '
+                'relationship.'
+            )
+
+        bucket_name = None
+        target = 'cloudify.aws.relationships.s3_object_contained_in_bucket'
+        for relationship in relationships:
+            if relationship.type == target:
+                bucket_name = relationship.target.node.properties['name']
+
+        if bucket_name is None:
+            raise NonRecoverableError(
+                'Could not get containing bucket name from related node.'
+            )
+
+        # If we got here, we should be dealing with a real bucket, so we'll
+        # just retrieve the permissions
+        bucket_permissions = self.execute(
+            self.client.get_bucket_acl,
+            {'Bucket': bucket_name},
+        )
+        bucket_permissions.pop('ResponseMetadata')
+
+        return bucket_name, bucket_permissions
+
+    def create(self, args=None):
+        bucket_name, bucket_permissions = self._get_bucket_details()
+
+        object_acl = ctx.node.properties.get('permissions', bucket_permissions)
+
+        # Get the contents var
+        is_file = False
+        if ctx.node.properties['load_contents_from_file']:
+            # It's a file, get a file handle for it
+            contents = ctx.download_resource(ctx.node.properties['contents'])
```

Comment:
I'm really not crazy about this. A node property should be an abstraction of the resource's static properties. On the other hand, there isn't a good 'Cloudify'-y way of doing things here. The most correct way would be to provide a separate node_template that represents a file object somewhere and to use it, but that's really hard from the user's perspective. Perhaps a good compromise would be a "contents" property and that property will be a new data type with a type field, where you specify "file" or "inline". What do you think?